In [1]:
import pandas as pd
import numpy as np
import matplotlib .pyplot as plt

Attribute Information:

InvoiceNo: Invoice number. Nominal. A 6-digit integral number uniquely assigned to each transaction. If this code starts with the letter 'c', it indicates a cancellation.
StockCode: Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.
Description: Product (item) name. Nominal.
Quantity: The quantities of each product (item) per transaction. Numeric.
InvoiceDate: Invice date and time. Numeric. The day and time when a transaction was generated.
UnitPrice: Unit price. Numeric. Product price per unit in sterling (£).
CustomerID: Customer number. Nominal. A 5-digit integral number uniquely assigned to each customer.
Country: Country name. Nominal. The name of the country where a customer resides.

In [2]:
df = pd.read_csv("../Data/Raw/online_retail.CSV")

DATA CLEANING---
1. exploring data
2. identifying duplicates and nulls
3. handling duplicates
4. handling null values
5. changing data types appropriately

1. exploring data

In [3]:
df.tail()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
1067366,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
1067367,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
1067368,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
1067369,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.0,France
1067370,581587,POST,POSTAGE,1,2011-12-09 12:50:00,18.00,12680.0,France


In [4]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [5]:
df.describe()

,Quantity,Price,Customer ID
count,1.067371e+06,1.067371e+06,824364.000000
mean,9.938898e+00,4.649388e+00,15324.638504
std,1.727058e+02,1.235531e+02,1697.464450
min,-8.099500e+04,-5.359436e+04,12346.000000
25%,1.000000e+00,1.250000e+00,13975.000000
50%,3.000000e+00,2.100000e+00,15255.000000
75%,1.000000e+01,4.150000e+00,16797.000000
max,8.099500e+04,3.897000e+04,18287.000000


In [6]:
df.shape

(1067371, 8)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


In [8]:
df.duplicated().sum()

np.int64(34335)

In [9]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

1.The shape of data is 1067371 rows & 8 columns

2.there are 34335 duplicate records 

3.decription has 4382 missing values and customer_id has 243007

4.we can clearly see that  InvoiceDate is not a datetype and customer id is float remaining data is in proper format

In [10]:
df_clean = df

In [11]:
df_clean.duplicated().sum()

np.int64(34335)

In [12]:
df_clean[df_clean.duplicated()]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
...,...,...,...,...,...,...,...,...
1067136,581538,22068,BLACK PIRATE TREASURE CHEST,1,2011-12-09 11:34:00,0.39,14446.0,United Kingdom
1067150,581538,23318,BOX OF 6 MINI VINTAGE CRACKERS,1,2011-12-09 11:34:00,2.49,14446.0,United Kingdom
1067153,581538,22992,REVOLVER WOODEN RULER,1,2011-12-09 11:34:00,1.95,14446.0,United Kingdom
1067160,581538,22694,WICKER STAR,1,2011-12-09 11:34:00,2.10,14446.0,United Kingdom


In [13]:
df_clean[df_clean.duplicated(keep = False)].sort_values(by=df_clean.columns.tolist())

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
...,...,...,...,...,...,...,...,...
965610,C574510,22360,GLASS JAR ENGLISH CONFECTIONERY,-1,2011-11-04 13:25:00,2.95,15110.0,United Kingdom
986868,C575940,23309,SET OF 60 I LOVE LONDON CAKE CASES,-24,2011-11-13 11:38:00,0.55,17838.0,United Kingdom
986869,C575940,23309,SET OF 60 I LOVE LONDON CAKE CASES,-24,2011-11-13 11:38:00,0.55,17838.0,United Kingdom
1055441,C580764,22667,RECIPE BOX RETROSPOT,-12,2011-12-06 10:38:00,2.95,14562.0,United Kingdom


In [14]:
df_clean.shape

(1067371, 8)

Duplicates here are rows identical across all 8 columns, but InvoiceDate is only precise to the minute — so genuine repeat scans of the same item (e.g. added twice instead of updating quantity) can look identical to an error. Since we can't prove these are mistakes, and dropping them would understate Revenue and Frequency, we keep them.

Flagging the invoice starting "C" as cancellation(True/False) with a helper column called "is_cancellation"

In [15]:
df_clean["is_cancellation"] = df_clean["Invoice"].astype(str).str.startswith("C")

In [16]:
df_clean[df_clean["is_cancellation"]]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,is_cancellation
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia,True
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia,True
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia,True
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia,True
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia,True
...,...,...,...,...,...,...,...,...,...
1065910,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397.0,United Kingdom,True
1067002,C581499,M,Manual,-1,2011-12-09 10:28:00,224.69,15498.0,United Kingdom,True
1067176,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311.0,United Kingdom,True
1067177,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315.0,United Kingdom,True


HANDLING MISSING VALUES

In [17]:
df_clean.isnull().sum()

Invoice                 0
StockCode               0
Description          4382
Quantity                0
InvoiceDate             0
Price                   0
Customer ID        243007
Country                 0
is_cancellation         0
dtype: int64

In [18]:
df_clean[df_clean["Customer ID"].isnull()]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,is_cancellation
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.00,NaN,United Kingdom,False
283,489463,71477,short,-240,2009-12-01 10:52:00,0.00,NaN,United Kingdom,False
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.00,NaN,United Kingdom,False
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.00,NaN,United Kingdom,False
577,489525,85226C,BLUE PULL BACK RACING CAR,1,2009-12-01 11:49:00,0.55,NaN,United Kingdom,False
...,...,...,...,...,...,...,...,...,...
1066997,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,NaN,United Kingdom,False
1066998,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,NaN,United Kingdom,False
1066999,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,NaN,United Kingdom,False
1067000,581498,85174,S/4 CACTI CANDLES,1,2011-12-09 10:26:00,10.79,NaN,United Kingdom,False


In [19]:
df_clean[df_clean["Description"].isnull() & df_clean["Customer ID"].isnull()]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,is_cancellation
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom,False
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom,False
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom,False
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom,False
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom,False
...,...,...,...,...,...,...,...,...,...
1060783,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom,False
1060787,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom,False
1060793,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom,False
1062442,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom,False


Here we can see the description is null where all the customer_id is null so we can drop all records where customer_id and description are null as this is customer analytics we move with deleting rows were customer_id is null

In [20]:
df_clean = df_clean.dropna(subset = ["Customer ID"])

In [21]:
df_clean.shape

(824364, 9)

After deleting the missing values from customer_id columns(243007)
the final shape is  824364 rows × 8 columns

we can clearly see that  InvoiceDate is not a datetype remaining data is in proper format

In [22]:
df_clean["InvoiceDate"] = pd.to_datetime(df_clean["InvoiceDate"])

C:\Users\katak\AppData\Local\Temp\ipykernel_17748\2779893561.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean["InvoiceDate"] = pd.to_datetime(df_clean["InvoiceDate"])


In [23]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 824364 entries, 0 to 1067370
Data columns (total 9 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   Invoice          824364 non-null  object        
 1   StockCode        824364 non-null  object        
 2   Description      824364 non-null  object        
 3   Quantity         824364 non-null  int64         
 4   InvoiceDate      824364 non-null  datetime64[ns]
 5   Price            824364 non-null  float64       
 6   Customer ID      824364 non-null  float64       
 7   Country          824364 non-null  object        
 8   is_cancellation  824364 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 57.4+ MB


In [24]:
df_clean.columns = [c.strip().replace(" ","_") for c in df_clean.columns]
df_clean.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer_ID', 'Country', 'is_cancellation'],
      dtype='object')

In [25]:
df_clean["Customer_ID"] = df_clean["Customer_ID"].astype(int)

C:\Users\katak\AppData\Local\Temp\ipykernel_17748\530127439.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean["Customer_ID"] = df_clean["Customer_ID"].astype(int)


In [26]:
df_clean.nunique()

Invoice            44876
StockCode           4646
Description         5299
Quantity             643
InvoiceDate        41439
Price               1022
Customer_ID         5942
Country               41
is_cancellation        2
dtype: int64

In [27]:
df_clean.dtypes

Invoice                    object
StockCode                  object
Description                object
Quantity                    int64
InvoiceDate        datetime64[ns]
Price                     float64
Customer_ID                 int64
Country                    object
is_cancellation              bool
dtype: object

19494 rows × 9 columns

In [28]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 824364 entries, 0 to 1067370
Data columns (total 9 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   Invoice          824364 non-null  object        
 1   StockCode        824364 non-null  object        
 2   Description      824364 non-null  object        
 3   Quantity         824364 non-null  int64         
 4   InvoiceDate      824364 non-null  datetime64[ns]
 5   Price            824364 non-null  float64       
 6   Customer_ID      824364 non-null  int64         
 7   Country          824364 non-null  object        
 8   is_cancellation  824364 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 57.4+ MB


In [29]:
df_clean["is_cancellation"].sum()

np.int64(18744)

In [30]:
(df_clean["Quantity"]<=0).sum()

np.int64(18744)

In [31]:
(df_clean["Price"]<=0).sum()

np.int64(71)

In [32]:
df_clean[df["Price"]<=0]

C:\Users\katak\AppData\Local\Temp\ipykernel_17748\2169767630.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_clean[df["Price"]<=0]


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer_ID,Country,is_cancellation
4674,489825,22076,6 RIBBONS EMPIRE,12,2009-12-02 13:34:00,0.0,16126,United Kingdom,False
6781,489998,48185,DOOR MAT FAIRY CAKE,2,2009-12-03 11:19:00,0.0,15658,United Kingdom,False
16107,490727,M,Manual,1,2009-12-07 16:38:00,0.0,17231,United Kingdom,False
18738,490961,22065,CHRISTMAS PUDDING TRINKET POT,1,2009-12-08 15:25:00,0.0,14108,United Kingdom,False
18739,490961,22142,CHRISTMAS CRAFT WHITE FAIRY,12,2009-12-08 15:25:00,0.0,14108,United Kingdom,False
...,...,...,...,...,...,...,...,...,...
1004540,577129,22464,HANGING METAL HEART LANTERN,4,2011-11-17 19:52:00,0.0,15602,United Kingdom,False
1005014,577168,M,Manual,1,2011-11-18 10:42:00,0.0,12603,Germany,False
1006110,577314,23407,SET OF 2 TRAYS HOME SWEET HOME,2,2011-11-18 13:23:00,0.0,12444,Norway,False
1011446,577696,M,Manual,1,2011-11-21 11:57:00,0.0,16406,United Kingdom,False


In [33]:
((df_clean["Price"]<=0) & (df_clean["is_cancellation"] == True)).sum()

np.int64(0)

Rows with Price <= 0 that are NOT cancellations are stock adjustments (damages, missing, found, etc.), 
not real customer purchases. Dropping these. Genuine cancellations (Invoice starts with 'C') are kept 
since they have real prices and represent actual customer returns.

In [34]:
no_price= df_clean[df["Price"]<=0]

C:\Users\katak\AppData\Local\Temp\ipykernel_17748\979508792.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  no_price= df_clean[df["Price"]<=0]


In [35]:
no_price

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer_ID,Country,is_cancellation
4674,489825,22076,6 RIBBONS EMPIRE,12,2009-12-02 13:34:00,0.0,16126,United Kingdom,False
6781,489998,48185,DOOR MAT FAIRY CAKE,2,2009-12-03 11:19:00,0.0,15658,United Kingdom,False
16107,490727,M,Manual,1,2009-12-07 16:38:00,0.0,17231,United Kingdom,False
18738,490961,22065,CHRISTMAS PUDDING TRINKET POT,1,2009-12-08 15:25:00,0.0,14108,United Kingdom,False
18739,490961,22142,CHRISTMAS CRAFT WHITE FAIRY,12,2009-12-08 15:25:00,0.0,14108,United Kingdom,False
...,...,...,...,...,...,...,...,...,...
1004540,577129,22464,HANGING METAL HEART LANTERN,4,2011-11-17 19:52:00,0.0,15602,United Kingdom,False
1005014,577168,M,Manual,1,2011-11-18 10:42:00,0.0,12603,Germany,False
1006110,577314,23407,SET OF 2 TRAYS HOME SWEET HOME,2,2011-11-18 13:23:00,0.0,12444,Norway,False
1011446,577696,M,Manual,1,2011-11-21 11:57:00,0.0,16406,United Kingdom,False


In [36]:
df_clean = df_clean[df_clean["Price"] > 0]

In [37]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 824293 entries, 0 to 1067370
Data columns (total 9 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   Invoice          824293 non-null  object        
 1   StockCode        824293 non-null  object        
 2   Description      824293 non-null  object        
 3   Quantity         824293 non-null  int64         
 4   InvoiceDate      824293 non-null  datetime64[ns]
 5   Price            824293 non-null  float64       
 6   Customer_ID      824293 non-null  int64         
 7   Country          824293 non-null  object        
 8   is_cancellation  824293 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 57.4+ MB


In [38]:
df_clean.shape

(824293, 9)

creating two copies from df_clean one for sales and another for cancellation

In [39]:
Cancellation = df_clean[df_clean["is_cancellation"]].copy()
Cancellation.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18744 entries, 178 to 1067178
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Invoice          18744 non-null  object        
 1   StockCode        18744 non-null  object        
 2   Description      18744 non-null  object        
 3   Quantity         18744 non-null  int64         
 4   InvoiceDate      18744 non-null  datetime64[ns]
 5   Price            18744 non-null  float64       
 6   Customer_ID      18744 non-null  int64         
 7   Country          18744 non-null  object        
 8   is_cancellation  18744 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 1.3+ MB


In [40]:
Sales = df_clean[~ df_clean["is_cancellation"]].copy()
Sales.info()

<class 'pandas.core.frame.DataFrame'>
Index: 805549 entries, 0 to 1067370
Data columns (total 9 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   Invoice          805549 non-null  object        
 1   StockCode        805549 non-null  object        
 2   Description      805549 non-null  object        
 3   Quantity         805549 non-null  int64         
 4   InvoiceDate      805549 non-null  datetime64[ns]
 5   Price            805549 non-null  float64       
 6   Customer_ID      805549 non-null  int64         
 7   Country          805549 non-null  object        
 8   is_cancellation  805549 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 56.1+ MB


In [41]:
Sales["Revenue"] = Sales["Quantity"] * Sales["Price"]
Sales[["Invoice", "StockCode", "Quantity", "Price", "Revenue"]].head()

,Invoice,StockCode,Quantity,Price,Revenue
0,489434,85048,12,6.95,83.4
1,489434,79323P,12,6.75,81.0
2,489434,79323W,12,6.75,81.0
3,489434,22041,48,2.10,100.8
4,489434,21232,24,1.25,30.0


Revenue = Quantity * Price, calculated only on `sales` (cancellations excluded from revenue totals, 
kept separately for later return-rate calculation).

lets export sales and cancellation to csv

In [42]:
Sales.to_csv("../Data/processed/cleaned_transactions.csv",index = False)
Cancellation.to_csv("../Data/processed/cancellations.csv",index = False)

RFM Feature engineering
it gives an idea on which customers matter most to the business and how should each customer treated diffrently

Metrics
Recency - days since last purchase( it tells about customer is still active or not)
Frequency - number of distinct orders (diffrentiates loya repeat,buyer vs one_time purchase)
monetary - total revenue (how much do they actually bring in)
tenure - days since first purchase(new vs long term)
AOV - average_order_value - they buy a little or lot per order
return rate - cancelled/total orders are they clean customer or one who buys and sends a lot back.



Recency: since this dataset ends in Dec 2011 rather than today, Recency is calculated relative to the last transaction date in the data (+1 day)

In [43]:
reference_date = Sales["InvoiceDate"].max() + pd.Timedelta(days =1)

In [44]:
RFM = Sales.groupby("Customer_ID").agg(
    Recency = ("InvoiceDate", lambda x: (reference_date - x.max()).days),
    Frequency = ("Invoice","nunique"),
    Monetary = ("Revenue", "sum"),
    First_purchase = ("InvoiceDate","min"),
    Last_purchase = ("InvoiceDate","max"),
    Total_items = ("Quantity","sum")
    ).reset_index()

In [45]:
RFM.head()

,Customer_ID,Recency,Frequency,Monetary,First_purchase,Last_purchase,Total_items
0,12346,326,12,77556.46,2009-12-14 08:34:00,2011-01-18 10:01:00,74285
1,12347,2,8,5633.32,2010-10-31 14:20:00,2011-12-07 15:52:00,3286
2,12348,75,5,2019.40,2010-09-27 14:59:00,2011-09-25 13:13:00,2714
3,12349,19,4,4428.69,2010-04-29 13:20:00,2011-11-21 09:51:00,1624
4,12350,310,1,334.40,2011-02-02 16:01:00,2011-02-02 16:01:00,197


Tenure — days since first purchase → new customer vs. long-term?
AOV — average order value → do they buy a little or a lot per order?

In [46]:
RFM["Tenure"] = (reference_date - RFM["First_purchase"]).dt.days
RFM["AOV"] = RFM["Monetary"] / RFM["Frequency"]

Return rate  cancellations weere kept seperately from sales 
Here we count each customers no.of of unique invoices from cancellations and later merge that with RFM Table and compute the return Rate 

RETURN_RATE = CANCELLED_ORDERS /  (frequency +  cancelled_orders)
 first we have to create cancell counts and then merge it with RFM

In [47]:
Cancel_counts = Cancellation.groupby("Customer_ID")["Invoice"].nunique().reset_index()

In [48]:
Cancel_counts.columns = ["Customer_ID","Cancelled_orders"]

In [49]:
RFM = RFM.merge(Cancel_counts, on = "Customer_ID", how = "left")

In [50]:
RFM.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5878 entries, 0 to 5877
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Customer_ID       5878 non-null   int64         
 1   Recency           5878 non-null   int64         
 2   Frequency         5878 non-null   int64         
 3   Monetary          5878 non-null   float64       
 4   First_purchase    5878 non-null   datetime64[ns]
 5   Last_purchase     5878 non-null   datetime64[ns]
 6   Total_items       5878 non-null   int64         
 7   Tenure            5878 non-null   int64         
 8   AOV               5878 non-null   float64       
 9   Cancelled_orders  2511 non-null   float64       
dtypes: datetime64[ns](2), float64(3), int64(5)
memory usage: 459.3 KB


In [51]:
RFM["Cancelled_orders"] = RFM["Cancelled_orders"].fillna(0)

In [52]:
RFM["Return_Rate"] = RFM["Cancelled_orders"] / (RFM["Frequency"] + RFM["Cancelled_orders"])

In [53]:
RFM.head()

,Customer_ID,Recency,Frequency,Monetary,First_purchase,Last_purchase,Total_items,Tenure,AOV,Cancelled_orders,Return_Rate
0,12346,326,12,77556.46,2009-12-14 08:34:00,2011-01-18 10:01:00,74285,726,6463.038333,5.0,0.294118
1,12347,2,8,5633.32,2010-10-31 14:20:00,2011-12-07 15:52:00,3286,404,704.165000,0.0,0.000000
2,12348,75,5,2019.40,2010-09-27 14:59:00,2011-09-25 13:13:00,2714,438,403.880000,0.0,0.000000
3,12349,19,4,4428.69,2010-04-29 13:20:00,2011-11-21 09:51:00,1624,589,1107.172500,1.0,0.200000
4,12350,310,1,334.40,2011-02-02 16:01:00,2011-02-02 16:01:00,197,310,334.400000,0.0,0.000000


In [54]:
print("RFM shape:", RFM.shape)
RFM[["Recency","Frequency","Monetary","Tenure","AOV","Return_Rate"]].describe()

RFM shape: (5878, 11)


,Recency,Frequency,Monetary,Tenure,AOV,Return_Rate
count,5878.000000,5878.000000,5878.000000,5878.000000,5878.000000,5878.000000
mean,201.331916,6.289384,3018.616737,474.711637,391.726538,0.117530
std,209.338707,13.009406,14737.731040,223.098342,1215.071054,0.165489
min,1.000000,1.000000,2.950000,1.000000,2.950000,0.000000
25%,26.000000,1.000000,348.762500,313.000000,181.652500,0.000000
50%,96.000000,3.000000,898.915000,530.000000,285.070500,0.000000
75%,380.000000,7.000000,2307.090000,668.000000,420.570000,0.218410
max,739.000000,398.000000,608821.650000,739.000000,84236.250000,0.800000


In [55]:
RFM.to_csv("../Data/processed/customer_features.csv",index = False)
print("Saved:", RFM.shape)

Saved: (5878, 11)


RFM Scoring
Ranks every customer into quintilies (1-5) on Recency, Frequency, Monetary then later on three scores we will segment customers with labels eg. champions at risk, lost numbers into groups marketimng can actually act on 
target high-value segments win back at risk ones stop spending on lost customers

In [56]:
RFM["R_Score"] = pd.qcut(RFM["Recency"], 5 , labels = [5,4,3,2,1]).astype(int)

In [57]:
RFM["F_Score"] = pd.qcut(RFM["Frequency"].rank(method = "first"),5, labels = [1,2,3,4,5]).astype(int)

In [58]:
RFM["M_Score"] = pd.qcut(RFM["Monetary"], 5 , labels = [1,2,3,4,5]).astype(int)

In [59]:
RFM

,Customer_ID,Recency,Frequency,Monetary,First_purchase,Last_purchase,Total_items,Tenure,AOV,Cancelled_orders,Return_Rate,R_Score,F_Score,M_Score
0,12346,326,12,77556.46,2009-12-14 08:34:00,2011-01-18 10:01:00,74285,726,6463.038333,5.0,0.294118,2,5,5
1,12347,2,8,5633.32,2010-10-31 14:20:00,2011-12-07 15:52:00,3286,404,704.165000,0.0,0.000000,5,4,5
2,12348,75,5,2019.40,2010-09-27 14:59:00,2011-09-25 13:13:00,2714,438,403.880000,0.0,0.000000,3,4,4
3,12349,19,4,4428.69,2010-04-29 13:20:00,2011-11-21 09:51:00,1624,589,1107.172500,1.0,0.200000,5,3,5
4,12350,310,1,334.40,2011-02-02 16:01:00,2011-02-02 16:01:00,197,310,334.400000,0.0,0.000000,2,1,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5873,18283,4,22,2736.65,2010-02-19 17:16:00,2011-12-06 12:02:00,1733,658,124.393182,0.0,0.000000,5,5,4
5874,18284,432,1,461.68,2010-10-04 11:33:00,2010-10-04 11:33:00,494,432,461.680000,1.0,0.500000,1,2,2
5875,18285,661,1,427.00,2010-02-17 10:24:00,2010-02-17 10:24:00,145,661,427.000000,0.0,0.000000,1,2,2
5876,18286,477,2,1296.43,2009-12-16 10:45:00,2010-08-20 11:57:00,608,724,648.215000,1.0,0.333333,1,3,4


In [60]:
RFM.describe()

,Customer_ID,Recency,Frequency,Monetary,First_purchase,Last_purchase,Total_items,Tenure,AOV,Cancelled_orders,Return_Rate,R_Score,F_Score,M_Score
count,5878.000000,5878.000000,5878.000000,5878.000000,5878,5878,5878.000000,5878.000000,5878.000000,5878.000000,5878.000000,5878.000000,5878.000000,5878.000000
mean,15315.313542,201.331916,6.289384,3018.616737,2010-08-22 06:57:33.297039616,2011-05-22 16:19:59.469207296,1821.396223,474.711637,391.726538,1.333106,0.117530,3.005104,3.000000,3.000000
min,12346.000000,1.000000,1.000000,2.950000,2009-12-01 07:45:00,2009-12-01 09:55:00,1.000000,1.000000,2.950000,0.000000,0.000000,1.000000,1.000000,1.000000
25%,13833.250000,26.000000,1.000000,348.762500,2010-02-09 14:01:15,2010-11-25 10:24:45,190.000000,313.000000,181.652500,0.000000,0.000000,2.000000,2.000000,2.000000
50%,15314.500000,96.000000,3.000000,898.915000,2010-06-27 13:31:30,2011-09-05 11:59:00,491.500000,530.000000,285.070500,0.000000,0.000000,3.000000,3.000000,3.000000
75%,16797.750000,380.000000,7.000000,2307.090000,2011-01-30 14:30:15,2011-11-14 11:31:15,1376.000000,668.000000,420.570000,1.000000,0.218410,4.000000,4.000000,4.000000
max,18287.000000,739.000000,398.000000,608821.650000,2011-12-09 12:16:00,2011-12-09 12:50:00,367193.000000,739.000000,84236.250000,112.000000,0.800000,5.000000,5.000000,5.000000
std,1715.572666,209.338707,13.009406,14737.731040,NaN,NaN,8963.011855,223.098342,1215.071054,3.672388,0.165489,1.416969,1.414454,1.414454


In [61]:
RFM.groupby("R_Score")["Customer_ID"].nunique()

R_Score
1    1175
2    1172
3    1167
4    1176
5    1188
Name: Customer_ID, dtype: int64

In [62]:
RFM.groupby("F_Score")["Customer_ID"].nunique()

F_Score
1    1176
2    1175
3    1176
4    1175
5    1176
Name: Customer_ID, dtype: int64

In [63]:
RFM.groupby("M_Score")["Customer_ID"].nunique()

M_Score
1    1176
2    1175
3    1176
4    1175
5    1176
Name: Customer_ID, dtype: int64

labelling the segments 
RFM Segment label
1. r>=4 , f>=4 , m>= 4 --- champions
2. r>=3 , f>=3 , m>= 3 --- Loyal customers
3. r>=4 , f<=2 , --- New_Customers
4. r<=2, f>=3, ---- at risk
5. r<=2, f<=2 and m<=2 ---- lost
6. anthing else needs attention

In [64]:
def segment_customer(x):
    r,f,m = x["R_Score"], x["F_Score"], x["M_Score"]
    if r>=4 and f>=4 and m>=4:
        return "Champions"
    elif r>=3 and f>=3 and m>=3:
        return "Loyal_customers"
    elif r>=4 and f<=2:
        return "New_customers"
    elif r<=2 and f>=3:
        return "At_Risk"
    elif r<=2 and f<=2 and m<=2:
        return "lost" 
    else:
        return "Needs_Attention"

In [65]:
RFM["Segment"] = RFM.apply(segment_customer,axis =1)

In [66]:
Segment_counts = RFM["Segment"].value_counts()

In [67]:
segment_pct = (Segment_counts/len(RFM)*100).round(1)
segment_pct

Segment
Champions          22.1
lost               21.7
Loyal_customers    19.3
Needs_Attention    15.3
At_Risk            14.0
New_customers       7.5
Name: count, dtype: float64

In [68]:
Segment_summary = Segment_counts.reset_index()
Segment_summary.columns = ["Segment","Customer_count"]
Segment_summary["Pct_customers"] = segment_pct.values
Segment_summary

,Segment,Customer_count,Pct_customers
0,Champions,1300,22.1
1,lost,1275,21.7
2,Loyal_customers,1134,19.3
3,Needs_Attention,902,15.3
4,At_Risk,824,14.0
5,New_customers,443,7.5


CLV- Customer_lifetime value

CLV per segment
it gives a clear idea on how much ongoing value does each segment bring to the company
RFM Already told us what the segments are and CLV adds revenue number behind each segment
approach
CLV = AOV * Purchase frequency *customer lifespan
we here already have AOV and Purchase frequency in RFM
LIfespan_years = tenure/365 


In [69]:
RFM["Lifespan_years"] = RFM["Tenure"]/365
RFM["Lifespan_years"] = RFM["Lifespan_years"].replace(0,1/365)

In [70]:
RFM["CLV"] = RFM["AOV"] * RFM["Frequency"] * RFM["Lifespan_years"]

In [71]:
RFM[["Customer_ID","AOV","Frequency", "Lifespan_years", "CLV"]].head(10)

,Customer_ID,AOV,Frequency,Lifespan_years,CLV
0,12346,6463.038333,12,1.989041,154262.986192
1,12347,704.165000,8,1.106849,6235.236384
2,12348,403.880000,5,1.200000,2423.280000
3,12349,1107.172500,4,1.613699,7146.570986
4,12350,334.400000,1,0.849315,284.010959
5,12351,300.930000,1,1.027397,309.174658
6,12352,284.984000,10,1.076712,3068.457863
7,12353,203.380000,2,1.120548,455.794082
8,12354,1079.400000,1,0.635616,686.084384
9,12355,473.805000,2,1.556164,1474.636932


Now lets check CLV Per segment (avg clv, total clv, customer count)

In [72]:
CLV_by_segment = RFM.groupby("Segment").agg(
    Avg_CLV = ("CLV","mean"),
    Total_CLV = ("CLV","sum"),
    Customer_count =("Customer_ID","count")
).sort_values("Avg_CLV",ascending = False).round(2)
CLV_by_segment

,Avg_CLV,Total_CLV,Customer_count
Segment,,,
Champions,16780.67,21814872.19,1300
At_Risk,3501.87,2885538.63,824
Loyal_customers,3461.48,3925323.41,1134
Needs_Attention,795.07,717156.73,902
New_customers,424.53,188066.10,443
lost,346.44,441716.79,1275


In [73]:
RFM.to_csv("../Data/processed/customer_features.csv",index =False)

In [74]:
RFM.shape
RFM.columns

Index(['Customer_ID', 'Recency', 'Frequency', 'Monetary', 'First_purchase',
       'Last_purchase', 'Total_items', 'Tenure', 'AOV', 'Cancelled_orders',
       'Return_Rate', 'R_Score', 'F_Score', 'M_Score', 'Segment',
       'Lifespan_years', 'CLV'],
      dtype='object')

Here we can Observe Champion(22%) brings 73% of customers base value

At risk and brings as muchvalue as Loyal_customers but total_clv is less than the loyal_customers due to less headcount

Protect champions aggresively cause they are caryying the majority value even though for smaller group losing few of them is very costly

At Risk is the highest-priority win-back target -worth retentiom spend

Lost and New Customers contribute the least its not worth heavy investment although new customers are early in their journey and could be nurtered towards loyal/champions status over time


Cohort / Retention Analysis

Once a customer joins how long do they actually keep buying, this tells are we retaining and customers overtime or losing most of them quickly after thier first purchase

RFM and CLV looked after customer and segmentation but here 
cohort groups customers by when they joined a cohort means how many customers joined or made thier first purchase in same month
then we will look after the percentage of customers are still active after first purchase

if say only 20% of customers are still active 3 months after thier first purchase thats a retention problem woth solving

process- 
cohort retention table - rows are month a customer first purchased
columns - months since that first purchase
values - % of that cohort still active

In [75]:
Sales["Order_Month"] = Sales["InvoiceDate"].dt.to_period("M")

In [76]:
cohort_map = Sales.groupby("Customer_ID")["Order_Month"].min().reset_index()
cohort_map.columns = ["Customer_ID","Cohort_Month"]

This is a separate, smaller table — one row per customer, holding just their Cohort_Month (the month of their very first purchase, computed by finding the minimum Order_Month across all their transactions).

In [77]:
## Lets merge sales with Sales so that we can have clear idea on To answer "how many months after joining did this transaction happen," you need both pieces of information on the same row:
Sales = Sales.merge(cohort_map, on="Customer_ID", how="left")

In [78]:
Sales[["Customer_ID","Cohort_Month","Order_Month"]].sort_values("Customer_ID", ascending = True)

,Customer_ID,Cohort_Month,Order_Month
192093,12346,2009-12,2010-06
192097,12346,2009-12,2010-06
192098,12346,2009-12,2010-06
192099,12346,2009-12,2010-06
192094,12346,2009-12,2010-06
...,...,...,...
150143,18287,2010-05,2010-05
150142,18287,2010-05,2010-05
150141,18287,2010-05,2010-05
150118,18287,2010-05,2010-05


In [79]:
Sales 

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer_ID,Country,is_cancellation,Revenue,Order_Month,Cohort_Month
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085,United Kingdom,False,83.40,2009-12,2009-12
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085,United Kingdom,False,81.00,2009-12,2009-12
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085,United Kingdom,False,81.00,2009-12,2009-12
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085,United Kingdom,False,100.80,2009-12,2009-12
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085,United Kingdom,False,30.00,2009-12,2009-12
...,...,...,...,...,...,...,...,...,...,...,...,...
805544,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680,France,False,12.60,2011-12,2011-08
805545,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680,France,False,16.60,2011-12,2011-08
805546,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680,France,False,16.60,2011-12,2011-08
805547,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680,France,False,14.85,2011-12,2011-08


In [80]:
Sales["Cohort_Index"] = ((Sales["Order_Month"].dt.year - Sales["Cohort_Month"].dt.year)*12 + (Sales["Order_Month"].dt.month - Sales["Cohort_Month"].dt.month))

In [82]:
Sales[["Customer_ID","Cohort_Month","Order_Month","Cohort_Index"]].head()

,Customer_ID,Cohort_Month,Order_Month,Cohort_Index
0,13085,2009-12,2009-12,0
1,13085,2009-12,2009-12,0
2,13085,2009-12,2009-12,0
3,13085,2009-12,2009-12,0
4,13085,2009-12,2009-12,0


Lets build a retention table

In [87]:
cohort_data = Sales.groupby(["Cohort_Month","Cohort_Index"])["Customer_ID"].nunique().reset_index()
cohort_pivot = cohort_data.pivot(index="Cohort_Month",  columns = "Cohort_Index",values = "Customer_ID")

In [90]:
cohort_pivot

Cohort_Index,0,1,2,3,4,5,6,7,8,9,...,15,16,17,18,19,20,21,22,23,24
Cohort_Month,,,,,,,,,,,,,,,,,,,,,
2009-12,955.0,337.0,319.0,406.0,363.0,343.0,360.0,327.0,321.0,346.0,...,289.0,251.0,289.0,270.0,248.0,244.0,301.0,291.0,389.0,188.0
2010-01,383.0,79.0,119.0,117.0,101.0,115.0,99.0,88.0,107.0,122.0,...,58.0,90.0,76.0,71.0,75.0,93.0,74.0,94.0,22.0,NaN
2010-02,374.0,89.0,84.0,109.0,92.0,75.0,72.0,107.0,95.0,103.0,...,75.0,60.0,61.0,54.0,86.0,86.0,61.0,22.0,NaN,NaN
2010-03,443.0,84.0,102.0,107.0,103.0,90.0,109.0,134.0,122.0,48.0,...,75.0,77.0,69.0,78.0,89.0,94.0,35.0,NaN,NaN,NaN
2010-04,294.0,57.0,57.0,48.0,54.0,66.0,81.0,77.0,31.0,32.0,...,46.0,41.0,44.0,53.0,66.0,17.0,NaN,NaN,NaN,NaN
2010-05,254.0,40.0,43.0,44.0,45.0,65.0,54.0,32.0,15.0,21.0,...,32.0,35.0,42.0,39.0,12.0,NaN,NaN,NaN,NaN,NaN
2010-06,270.0,47.0,51.0,55.0,62.0,77.0,34.0,24.0,22.0,32.0,...,33.0,36.0,55.0,14.0,NaN,NaN,NaN,NaN,NaN,NaN
2010-07,186.0,29.0,34.0,55.0,54.0,26.0,21.0,27.0,27.0,21.0,...,32.0,44.0,15.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2010-08,162.0,33.0,48.0,52.0,28.0,19.0,16.0,20.0,22.0,21.0,...,32.0,11.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [92]:
cohort_size = cohort_pivot.iloc[:,0]

In [94]:
retention_table  = cohort_pivot.divide(cohort_size,axis  = 0).round(3)*100

In [95]:
retention_table

Cohort_Index,0,1,2,3,4,5,6,7,8,9,...,15,16,17,18,19,20,21,22,23,24
Cohort_Month,,,,,,,,,,,,,,,,,,,,,
2009-12,100.0,35.3,33.4,42.5,38.0,35.9,37.7,34.2,33.6,36.2,...,30.3,26.3,30.3,28.3,26.0,25.5,31.5,30.5,40.7,19.7
2010-01,100.0,20.6,31.1,30.5,26.4,30.0,25.8,23.0,27.9,31.9,...,15.1,23.5,19.8,18.5,19.6,24.3,19.3,24.5,5.7,NaN
2010-02,100.0,23.8,22.5,29.1,24.6,20.1,19.3,28.6,25.4,27.5,...,20.1,16.0,16.3,14.4,23.0,23.0,16.3,5.9,NaN,NaN
2010-03,100.0,19.0,23.0,24.2,23.3,20.3,24.6,30.2,27.5,10.8,...,16.9,17.4,15.6,17.6,20.1,21.2,7.9,NaN,NaN,NaN
2010-04,100.0,19.4,19.4,16.3,18.4,22.4,27.6,26.2,10.5,10.9,...,15.6,13.9,15.0,18.0,22.4,5.8,NaN,NaN,NaN,NaN
2010-05,100.0,15.7,16.9,17.3,17.7,25.6,21.3,12.6,5.9,8.3,...,12.6,13.8,16.5,15.4,4.7,NaN,NaN,NaN,NaN,NaN
2010-06,100.0,17.4,18.9,20.4,23.0,28.5,12.6,8.9,8.1,11.9,...,12.2,13.3,20.4,5.2,NaN,NaN,NaN,NaN,NaN,NaN
2010-07,100.0,15.6,18.3,29.6,29.0,14.0,11.3,14.5,14.5,11.3,...,17.2,23.7,8.1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2010-08,100.0,20.4,29.6,32.1,17.3,11.7,9.9,12.3,13.6,13.0,...,19.8,6.8,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [96]:
retention_table.to_csv("../Data/processed/retention_table_for_dashboard.csv")


In [98]:
Segment_summary.to_csv("../outputs/segment_summary.csv", index=False)
CLV_by_segment.to_csv("../outputs/clv_by_segment.csv")
retention_table.to_csv("../outputs/cohort_retention.csv")